src/Arcus.WebApi.Tests.Unit/Logging/RequestTrackingMiddlewareTests.cs

```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Arcus.Observability.Telemetry.Core;
+using Arcus.WebApi.Logging;
+using Arcus.WebApi.Tests.Unit.Correlation;
+using Arcus.WebApi.Tests.Unit.Hosting;
+using Microsoft.AspNetCore.Builder;
+using Serilog.Events;
+using Xunit;
+
+namespace Arcus.WebApi.Tests.Unit.Logging
+{
+    public class RequestTrackingMiddlewareTests : IDisposable
+    {
+        private readonly TestApiServer _testServer = new TestApiServer();
+
+        [Fact]
+        public async Task GetRequest_TracksRequest_ReturnsSuccess()
+        {
+            // Arrange
+            const string headerName = "x-custom-header", headerValue = "custom header value";
+            _testServer.AddConfigure(app => app.UseMiddleware<RequestTrackingMiddleware>());
+            using (HttpClient client = _testServer.CreateClient())
+            {
+                var request = new HttpRequestMessage(HttpMethod.Get, EchoController.Route)
+                {
+                    Headers = { {headerName, headerValue} },
+                    Content = new StringContent("\"echo me\"", Encoding.UTF8, "application/json")
+                };
+
+                // Act
+                using (HttpResponseMessage response = await client.SendAsync(request))
+                {
+                    // Assert
+                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                    string content = await response.Content.ReadAsStringAsync();
+                    Assert.Equal("echo me", content);
+
+                    IReadOnlyDictionary<ScalarValue, LogEventPropertyValue> eventContext = GetLoggedEventContext();
+                    Assert.Contains(eventContext, item => item.Key.ToStringValue() == headerName && item.Value.ToStringValue() == $"[\"{headerValue}\"]");
+                    Assert.Contains(eventContext, item => item.Key.ToStringValue() == "Body" && item.Value.ToStringValue().Trim('\\', '\"') == "echo me");
+                }
+            }
+        }
+
+        [Fact]
+        public async Task GetRequestWithoutHeaders_TracksRequest_ReturnsSuccess()
+        {
+            // Arrange
+            const string headerName = "x-custom-header", headerValue = "custom header value", body = "echo me";
+            _testServer.AddConfigure(app => app.UseMiddleware<NoHeadersRequestTrackingMiddleware>());
```

Comment:
I'm not sure what we are trying to test here? Can you elaborate on the scenario please?

Reply:
Yes, the added `.AddConfigure()` method is a way to provide additional middleware components from the test. In this case, we want to verify if we can override the existing middleware and provide one of our own (in this case `NoHeadersRequestTrackingMiddleware`) and that indeed the "request headers are not" in the request tracking.

Reply:
The three tests are the three useful combinations of the available overridables. We could add also where both the headers and body is not present though.